### PR TITLE
test(e2e): add SBOM Groups BDD test coverage

### PR DESCRIPTION
**Pull request by mrrajan**
## Summary

- 38 BDD scenarios covering SBOM Groups UI (up from 14 existing): product labels, hierarchical tree, parent group selection, invalid group ID, SBOM counts, product badge, edit/remove parent, breadcrumbs, sorting, pagination, 3-level hierarchy, delete guards, and filter-by-parent
- 66 API tests across 11 suites: CRUD, ETag/optimistic locking, sorting, filtering/pagination, hierarchy, error cases, assignments (PUT/PATCH/bulk), and edge cases
- New page objects: `SbomGroupDetailPage`, `AddToGroupModal`; enhanced: `SbomGroupListPage`, `GroupFormModal`, `DetailsPageLayout`, `ConfirmDialog`
- API test helpers module: `sbom-group-helpers.ts`
- Review feedback fixes: removed explicit timeouts from shared page objects (TC-5189), separated filter-existence check from `clearAllFilters` (TC-5190)

## UI Scenarios Added

| # | Scenario | Priority |
|---|---|---|
| 1 | Product label badge appears for product groups | High |
| 2 | Expand and collapse hierarchical tree nodes | High |
| 3 | Create child group with parent selection | High |
| 4 | Navigate to invalid group ID shows error | Medium |
| 5 | SBOM count is displayed for groups with SBOMs | Medium |
| 6 | Product badge is displayed on group detail page | Medium |
| 7 | Edit group to assign a parent | Medium |
| 8 | Edit group to remove parent makes it a root group | Medium |
| 9 | Breadcrumb navigation on group detail page | Low |
| 10 | Pagination and sorting on group detail page | Low |
| 11 | Filter matching child auto-expands parent | High |
| 12 | Toggle product flag from Yes to No | Medium |
| 13 | SBOM count decreases when SBOM is reassigned | Medium |
| 14 | SBOM counts are independent between parent and child | Medium |
| 15 | Expand and navigate 3-level hierarchy | Low |
| 16 | Navigate into grandchild from 3-level tree | Low |
| 17 | Remove parent from middle group in 3-level hierarchy | Low |
| 18 | Delete group with SBOMs reassigns to parent | Medium |

## API Test Suites

- **CRUD** (16 tests): create, read, update, delete with name/description/labels/parent
- **ETag / optimistic locking** (2 tests): stale ETag rejection
- **Sorting** (2 tests): name ascending/descending
- **Filtering & pagination** (6 tests): name filter, label filter, offset/limit
- **Hierarchy** (5 tests): parent-child, multi-level, orphan on parent delete
- **Error cases** (7 tests): duplicate name, missing name, not found, invalid parent
- **Assignments** (28 tests): PUT replace, PATCH add/remove/clear, bulk assign, edge cases

## Not Implemented (features not yet in codebase)

- **Remove SBOM from group** — no UI exists for this action
- **"Groups" column on All SBOMs page** — column does not exist in SBOM table
- **Loading states** — requires component-level test, not E2E
- **React Query caching** — requires unit/integration test, not E2E

## Test plan

- [ ] Run `npm run e2e:test:ui` to verify all UI scenarios pass
- [ ] Run `npm run e2e:test:api` to verify all API tests pass
- [ ] Verify new scenarios follow BDD standards

Implements [TC-3811](https://redhat.atlassian.net/browse/TC-3811)

[TC-3811]: https://redhat.atlassian.net/browse/TC-3811?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: e2e/tests/ui/features/@sbom-groups/sbom-groups.step.ts [hunk 1]
```diff
@@ -462,3 +459,251 @@ When(
...
+// Invalid group ID handling
+When("User navigates to group details with invalid ID", async ({ page }) => {
+  await page.goto("/sbom-groups/invalid-group-id-12345");
+});
+
+Then("An error state is displayed for the invalid group", async ({ page }) => {
+  const errorHeading = page.getByRole("heading", {
+    name: /error|not found|something went wrong/i,
+  });
+  await expect(errorHeading).toBeVisible({ timeout: 10000 });
```
[the remark this answers is not in the document]
Comment (mrrajan) [reply e2e/tests/ui/features/@sbom-groups/sbom-groups.step.ts]: [sdlc-workflow/verify-pr] Classified as **suggestion** — proposes using data-testid selectors instead of heading regex; no documented convention in CONVENTIONS.md and data-testid is not an established E2E pattern in this repo (0 occurrences in e2e/). No sub-task created.

File: e2e/tests/ui/features/@sbom-groups/sbom-groups.step.ts [hunk 2]
```diff
@@ -447,7 +445,6 @@ When(
 Then(
```
[the remark this answers is not in the document]
Comment (mrrajan) [reply e2e/tests/ui/features/@sbom-groups/sbom-groups.step.ts:340]: [sdlc-workflow/verify-pr] Classified as **suggestion** — proposes parsing SBOM count to assert > 0; no documented assertion specificity convention in CONVENTIONS.md. The current implementation uses exact text matching which is functionally correct. No sub-task created.

File: e2e/tests/api/features/sbom-groups.ts [hunk 3]
```diff
@@ -0,0 +1,1118 @@
...
+  test.afterEach(async ({ axios }) => {
+    await cleanupGroups(axios, createdGroupIds.splice(0));
+  });
+
+  const findFirstSbomId = async (axios: import("axios").AxiosInstance) => {
+    const response = await axios.get("/api/v3/sbom", {
+      params: { limit: 1, offset: 0 },
+    });
+    expect(response.data.items.length).toBeGreaterThan(0);
+    return response.data.items[0].id as string;
```
[the remark this answers is not in the document]
Comment (mrrajan) [reply e2e/tests/api/features/sbom-groups.ts]: [sdlc-workflow/verify-pr] Classified as **suggestion** — proposes deduplicating findFirstSbomId/findTwoSbomIds helpers; valid code quality observation but no documented DRY test helper convention in CONVENTIONS.md to warrant upgrade. No sub-task created.

File: e2e/tests/api/features/sbom-groups.ts [hunk 4]
```diff
@@ -0,0 +1,1118 @@
...
+  });
+});
+
+// Added Patch tests - disabled for now
+// test.describe("SBOM Group PATCH assignments", () => {
+//   const createdGroupIds: string[] = [];
+
+//   test.afterEach(async ({ axios }) => {
+//     await cleanupGroups(axios, createdGroupIds.splice(0));
+//   });
```
[the remark this answers is not in the document]
Comment (mrrajan) [reply e2e/tests/api/features/sbom-groups.ts]: [sdlc-workflow/verify-pr] Classified as **question** — asks for clarification about commented-out PATCH assignment tests; no code change needed. No sub-task created.

**Comment by mrrajan** [comment]
[sdlc-workflow/verify-pr] Re: @sourcery-ai[bot] review (4569251325) — Classified as **suggestion** (3 items):
1. Extract shared helper utilities for setup logic — no documented DRY test helper convention in CONVENTIONS.md
2. Tighten error assertion regex — no documented testing selector convention
3. Use stable attributes/test-ids over text patterns — data-testid is not an established E2E pattern in this repo (0 occurrences in e2e/)

No sub-tasks created.

**Comment by mrrajan** [comment]
[sdlc-workflow/verify-pr] Re: @sourcery-ai[bot] review (4665095148) — Classified as **suggestion** (2 items):
1. Rename hasLabelBadge/hasSbomCount methods (return Locator, not boolean) — naming inconsistency is real but no documented method-naming convention for page objects in CONVENTIONS.md
2. Hardcoded #bulk-selected-items-checkbox ID — no documented convention about avoiding hardcoded ID selectors in tests

No sub-tasks created.

**Comment by mrrajan** [comment]
## Verification Report for TC-3811 (commit 3003ccf)

| Check | Result | Details |
|-------|--------|---------|
| Review Feedback | PASS | 6 items classified (4 suggestions, 1 question, 1 multi-suggestion review body); no code change requests |
| Root-Cause Investigation | N/A | No sub-tasks created |
| Scope Containment | WARN | All 15 files within e2e/; files beyond @sbom-groups/ are justified page objects, shared infrastructure, and test cleanup. User confirmed PR intentionally exceeds Jira scope |
| Diff Size | PASS | ~2,614 added / ~336 removed across 15 files — proportionate for ~17 test scenarios + API tests + page objects |
| Commit Traceability | WARN | 1/9 commits explicitly references TC-3811 (85385cc1); remaining 8 have descriptive messages but lack Jira ID. Traceability maintained at PR level |
| Sensitive Patterns | PASS | No secrets, credentials, or API keys detected in any added lines |
| CI Status | PASS | All 9 required checks pass; e2e-integration-tests passed in 10m20s |
| Acceptance Criteria | PASS | All 13 feasible e2e items covered; 4 items correctly excluded (no UI exists or out of e2e scope); PR includes substantial additional API test coverage |
| Test Quality | WARN | Repetitive test detection: duplicate findFirstSbomId/findTwoSbomIds helpers and repetitive try/catch error patterns in API tests. Test documentation: PASS (BDD scenarios + descriptive test names). Eval Quality: N/A |
| Test Change Classification | ADDITIVE | All files new or net-additive (+2,613/-322 lines); removed lines are refactored code, not removed coverage |
| Verification Commands | N/A | No verification commands specified |

### Overall: WARN

**Summary:** The PR substantially exceeds the Jira task requirements — adding 11 BDD scenarios, comprehensive API test coverage, and page object improvements. All CI checks pass and all feasible acceptance criteria are satisfied. The WARN verdicts are minor:
- **Scope Containment WARN**: Expected — user confirmed the PR intentionally includes API tests and infrastructure beyond the Jira's listed scenarios.
- **Commit Traceability WARN**: Only the initial commit references TC-3811; follow-up commits lack the Jira ID but are on the TC-3811 branch.
- **Test Quality WARN**: Duplicate helper functions (`findFirstSbomId`/`findTwoSbomIds`) and repetitive error-handling try/catch blocks could be extracted but do not affect correctness.

None of the reviewer suggestions (from sourcery-ai[bot]) were upgraded to code change requests — no matching CONVENTIONS.md patterns or established codebase conventions were found.

---
*This comment was AI-generated by [sdlc-workflow/verify-pr](https://github.com/mrizzi/sdlc-plugins) v0.11.0.*

File: e2e/tests/ui/pages/Pagination.ts [hunk 5]
```diff
@@ -43,7 +43,9 @@ export class Pagination {
       .getByRole("menuitem", { name: `${perPage} per page` })
       .click();
 
-    await expect(this._pagination.locator("input")).toHaveValue("1");
+    await expect(this._pagination.locator("input")).toHaveValue("1", {
+      timeout: 5000,
+    });
```
Comment (carlosthe19916) [inline e2e/tests/ui/pages/Pagination.ts]: Why is it that we need to add timeout here?
The whole Pagination class is independent of the any asynchronous call like REST APIs.
If changing the page triggers a Table or any other component to be re-rendered then the "wait/timeout" action should be written in the Table but not here.

File: e2e/tests/ui/pages/Toolbar.ts [hunk 6]
```diff
@@ -166,17 +166,20 @@ export class Toolbar<
...
-   * Clears all applied filters by clicking the "Clear all filters" button
+   * Clears all applied filters by clicking the "Clear all filters" button.
+   * No-ops gracefully when no filters are active (button not present).
    */
   async clearAllFilters() {
     const clearButton = this._toolbar.getByRole("button", {
       name: "Clear all filters",
     });
-    await expect(clearButton).toBeVisible();
+    if (!(await clearButton.isVisible())) return;
```
Comment (carlosthe19916) [inline e2e/tests/ui/pages/Toolbar.ts]: Let's not mix logic. The function should clear filters, verifying whether or not the filters exist should be done outside this function. Prefer composition logic over conditional expansion

File: e2e/tests/ui/pages/Toolbar.ts [hunk 7]
```diff
@@ -166,17 +166,20 @@ export class Toolbar<
...
     await clearButton.click();
 
-    // Verify all filter chips are removed
-    await expect(this._toolbar.locator(".pf-m-label-group")).toHaveCount(0);
+    // Give PatternFly label-group animations time to fully complete
+    await expect(this._toolbar.locator(".pf-m-label-group")).toHaveCount(0, {
+      timeout: 10000,
+    });
```
Comment (carlosthe19916) [inline e2e/tests/ui/pages/Toolbar.ts]: Don't add this. The whole Toolbar class is independent of any asynchronous event. If the toolbar triggers other components to be re-rendered asynchronously then the "wait" or async logic should be there not here

**Review by carlosthe19916 — CHANGES_REQUESTED** [review]
Do not use: "timeout" statements, avoid explicit timeout-based waits because they make tests slower, less reliable, and dependent on machine performance. Playwright already provides automatic waiting for elements, navigation, assertions, and actions, which makes tests more deterministic and less flaky.

File: e2e/tests/ui/features/@sbom-groups/sbom-groups.step.ts [hunk 8]
```diff
@@ -1,166 +1,90 @@
...
+  "Alert message {string} is displayed",
+  async ({ page }, message: string) => {
+    await expect(page.getByText(message)).toBeVisible();
   },
 );
 
-// Edit group
+// Given setup steps — group existence with POM
 Given("A group {string} exists", async ({ page }, groupName: string) => {
   const listPage = await SbomGroupListPage.fromCurrentPage(page);
```
[the remark this answers is not in the document]
Comment (mrrajan) [reply e2e/tests/ui/features/@sbom-groups/sbom-groups.step.ts:72]: [sdlc-workflow/verify-pr] Classified as **suggestion** — proposes strengthening the "table does not contain" assertion; no documented assertion pattern convention in CONVENTIONS.md and current implementation is functionally correct. No sub-task created.

File: e2e/tests/ui/pages/Pagination.ts (hunk 5 above)
Comment (mrrajan) [reply e2e/tests/ui/pages/Pagination.ts]: [sdlc-workflow/verify-pr] Classified as **code change request** — sub-task [TC-5189](https://redhat.atlassian.net/browse/TC-5189) created to address this feedback.

File: e2e/tests/ui/pages/Toolbar.ts (hunk 6 above)
Comment (mrrajan) [reply e2e/tests/ui/pages/Toolbar.ts]: [sdlc-workflow/verify-pr] Classified as **code change request** — sub-task [TC-5190](https://redhat.atlassian.net/browse/TC-5190) created to address this feedback.

File: e2e/tests/ui/pages/Toolbar.ts (hunk 7 above)
Comment (mrrajan) [reply e2e/tests/ui/pages/Toolbar.ts]: [sdlc-workflow/verify-pr] Classified as **code change request** — sub-task [TC-5189](https://redhat.atlassian.net/browse/TC-5189) created to address this feedback.

**Comment by mrrajan** [comment]
[sdlc-workflow/verify-pr] Re: @sourcery-ai[bot] review (4670363250) — Classified as **suggestion** (2 items):
1. Async methods returning Locator synchronously — valid observation but no documented async/sync naming convention for page objects in CONVENTIONS.md
2. Hardcoded SBOM names — no convention prohibiting hardcoded test data in Gherkin feature files

No sub-tasks created.

**Comment by mrrajan** [comment]
[sdlc-workflow/verify-pr] Re: @carlosthe19916 review (4682471904) — Classified as **code change request** — sub-task [TC-5189](https://redhat.atlassian.net/browse/TC-5189) created to remove explicit timeout waits from Pagination and Toolbar page objects.

[TC-5189]: https://redhat.atlassian.net/browse/TC-5189?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by mrrajan** [comment]
## Verification Report for TC-3811 (commit c0d2035)

| Check | Result | Details |
|-------|--------|---------|
| Review Feedback | WARN | 4 code change requests from maintainer; sub-tasks TC-5189, TC-5190 created |
| Root-Cause Investigation | DONE | Convention gap identified; TC-5191 created to document Playwright auto-waiting and page object design conventions |
| Scope Containment | WARN | All 16 files in e2e/; files outside @sbom-groups/ are justified shared infrastructure and page objects |
| Diff Size | PASS | +2349/-292 across 16 files, proportionate for 11+ UI scenarios + API test suite |
| Commit Traceability | WARN | 1 of 11 commits references TC-3811; remaining use generic prefixes |
| Sensitive Patterns | PASS | No secrets detected in 2349 added lines |
| CI Status | PASS | All 11 required checks pass; 7 conditional jobs skipped |
| Acceptance Criteria | PASS | 9 of 9 criteria met — product labels, tree hierarchy, parent selection, invalid ID, SBOM count, breadcrumbs, edit parent, sorting, API CRUD |
| Test Quality | WARN | Repetitive Test Detection: 18 identical try/catch blocks in API tests; Test Documentation: PASS; Eval Quality: N/A |
| Test Change Classification | MIXED | +25 scenarios, +34 step defs; 2 scenarios removed, ~10 step defs replaced with POM-based implementations |
| Verification Commands | N/A | None specified |

### Overall: WARN

**Issues requiring attention:**
1. **Timeouts in shared page objects** (TC-5189): Remove explicit `timeout` params from Pagination.ts and Toolbar.ts — Playwright auto-waiting handles this
2. **Mixed logic in clearAllFilters** (TC-5190): Separate filter-existence check from clear action in Toolbar.ts
3. **Commit traceability**: Only 1 of 11 commits references TC-3811; consider including the Jira key in all commit messages
4. **Repetitive test pattern**: 18 identical try/catch error assertion blocks in API tests could be extracted into a shared `expectHttpError()` helper

---
*This comment was AI-generated by [sdlc-workflow/verify-pr](https://github.com/RHEcosystemAppEng/sdlc-plugins) v0.13.2.*

**Comment by mrrajan** [comment]
@carlosthe19916 Fixed the mentioned issues and raised [PR#1154](https://github.com/guacsec/trustify-ui/pull/1154) for convention file update.

File: e2e/tests/ui/pages/Toolbar.ts [hunk 9]
```diff
@@ -166,16 +166,14 @@ export class Toolbar<
...
    */
   async clearAllFilters() {
     const clearButton = this._toolbar.getByRole("button", {
       name: "Clear all filters",
     });
-    await expect(clearButton).toBeVisible();
     await clearButton.click();
 
-    // Verify all filter chips are removed
     await expect(this._toolbar.locator(".pf-m-label-group")).toHaveCount(0);
```
Comment (carlosthe19916) [inline e2e/tests/ui/pages/Toolbar.ts:179]: I don't think there is a need to change what we have in the main branch. Just avoid changes.

**Review by carlosthe19916 — CHANGES_REQUESTED** [review]
- CI is not passing, always verify CI passes before any review is requested

Is there a way to avoid all the noise in the comments? 

- For each comment I added there was an AI generated reply that doesn't actually reply or add any info, it only says something like `sub-task created to address this feedback.` 
- There was a direct tag message 
```
[sdlc-workflow/verify-pr] Re: @carlosthe19916 review (4682471904) — Classified as code change request — sub-task TC-5189 created to remove explicit timeout waits from Pagination and Toolbar page objects.
```
That kind of comments only increase the noise both in github and emails (where notifications come)

**Comment by mrrajan** [comment]
> * CI is not passing, always verify CI passes before any review is requested
> 
> Is there a way to avoid all the noise in the comments?
> 
> * For each comment I added there was an AI generated reply that doesn't actually reply or add any info, it only says something like `sub-task created to address this feedback.`
> * There was a direct tag message
> 
> ```
> [sdlc-workflow/verify-pr] Re: @carlosthe19916 review (4682471904) — Classified as code change request — sub-task TC-5189 created to remove explicit timeout waits from Pagination and Toolbar page objects.
> ```
> 
> That kind of comments only increase the noise both in github and emails (where notifications come)

@carlosthe19916 I am currently using the Agentic SDLC workflow to help verify this PR and resolve the issue. The AI comments were automatically generated to track your requested changes as sub-tasks on JIRA. I will see if we can mute those automated workflow replies, this will evolve as we continue to refine our AI tooling and workflow integration

File: e2e/tests/ui/pages/Toolbar.ts (hunk 9 above)
Comment (mrrajan) [reply e2e/tests/ui/pages/Toolbar.ts:179]: @carlosthe19916 
> Let's not mix logic. The function should clear filters, verifying whether or not the filters exist should be done outside this function. Prefer composition logic over conditional expansion

Based on the earlier review comments, the CONVENTIONS.md file has been updated with [PR#1154](https://github.com/guacsec/trustify-ui/pull/1154/changes). And this is been removed as part of our coding convention. And the next step will fail if the button is not visible irrespective of the assertion. So, Please let me know what you think.

Comment (carlosthe19916) [reply e2e/tests/ui/pages/Toolbar.ts:179]: Please leave the code as it is in the main branch.

In my original review I suggested:
- Not to use `timeouts`
  - The code of the main branch does not use `timeout` statements.
- Not to add conditional logic and instead use composition.
  - The code of the main branch does not use conditional `if-else` to change the behavior of the function. It will always behave the same.
  - The original code of this PR initially did something like `if "filters exists then clean filters" else "do not clear filters and assume they don't exists"`. This is the conditional behavior we should avoid. 
    - this is the original code I asked to change `if (!(await clearButton.isVisible())) return;`

Let me expand of the "conditional" aspect: let's say you have a function "clickOnButton(id: buttonId)". The function's name is to "click".

What is wrong:
```
clickOnButton(id: buttonId) {
  if "button exists" then "click"
  else "don't click or do nothing"
}
```

What should be done:
```
clickOnButton(id: buttonId) {
  "click on button"
}

// Code
const button = getButton();
assert button.exists();
clickOnButton(button);
```

My apologies if my first comment was not clear enough, my bad

Comment (mrrajan) [reply e2e/tests/ui/pages/Toolbar.ts:179]: Thanks for the clarity. Totally fine with removing the timeout.

For the `clearAllFilters()`, your earlier point makes total sense - lets not mix actions and assertions. Playwright `click()` event already does actionalibility [checks](https://playwright.dev/docs/actionability) to make sure the element is visible. 
So, the element visibility check is redundant here:
```
await expect(clearButton).toBeVisible();
```

Referring to [PR#1154](https://github.com/guacsec/trustify-ui/pull/1154), I think leaving the main branch as it is would be inconsistent with the convention. Aligning the code with it will help both engineers and AI agents apply these patterns consistently.

That said, if you prefer to keep this check, I am good with to do so. In that case, I will update PR #1154 to remove the convention so the guidance stays consistent with the existing code.

Comment (carlosthe19916) [reply e2e/tests/ui/pages/Toolbar.ts:179]: please leave the current main code in regards of the function `clearAllFilters`

Comment (mrrajan) [reply e2e/tests/ui/pages/Toolbar.ts:179]: Sure. I will leave the `toBeVisible()` line in `clearAllFilters` for now so we can merge. 

Just wanted to note that it deviates from the convention of letting Playwright handle actionability checks. We should definitely get these patterns officially documented soon to keep both engineers and AI agents aligned.

**Review by carlosthe19916 — APPROVED** [review]
